packages/core/src/sagas/connectSaga.js

```diff
@@ -26,108 +25,155 @@ import {
 
-    yield take(DISCONNECT);
-    yield call(disconnectSaga, connectTask, directLine);
+    return connectionStatus !== CONNECTING && connectionStatus !== ONLINE;
   }
 }
 
-function* connectSaga(directLine, userID) {
-  const meta = { userID };
+function rectifyUserID(directLine, userIDFromAction) {
+  const { token } = directLine;
+  const { user: userIDFromToken } = decode(token) || {};
+
+  if (userIDFromToken) {
+    if (userIDFromAction && userIDFromAction !== userIDFromToken) {
+      console.warn('Web Chat: user ID is both specified in the Direct Line token and passed in, will use the user ID from the token.');
+    }
 
-  yield put({ type: CONNECT_PENDING, meta });
+    return userIDFromToken;
+  } else if (userIDFromAction) {
+    if (typeof userIDFromAction !== 'string') {
+      console.warn('Web Chat: user ID must be a string.');
 
-  const connectionStatusQueue = createPromiseQueue();
-  const connectionStatusSubscription = directLine.connectionStatus$.subscribe({ next: connectionStatusQueue.push });
+      return randomUserID();
+    } else if (/^dl_/.test(userIDFromAction)) {
+      console.warn('Web Chat: user ID prefixed with "dl_" is reserved and must be embedded into the Direct Line token to prevent forgery.');
+
+      return randomUserID();
+    }
+  } else {
+    return randomUserID();
+  }
 
+  return userIDFromAction;
+}
+
+function* connectSaga(directLine) {
   // DirectLineJS start the connection only after the first subscriber for activity$, but not connectionStatus$
   const activitySubscription = directLine.activity$.subscribe({ next: () => 0 });
 
   try {
-    try {
-      yield callUntil(connectionStatusQueue.shift, [], connectionStatus => connectionStatus === ONLINE);
-      yield put({ type: CONNECT_FULFILLING, meta, payload: { directLine } });
-      yield put({ type: CONNECT_FULFILLED, meta, payload: { directLine } });
-    } catch (err) {
-      yield put({ type: CONNECT_REJECTED, error: true, meta, payload: err });
-    } finally {
-      if (yield cancelled()) {
-        yield put({ type: CONNECT_REJECTED, error: true, meta, payload: new Error('cancelled') });
+    for (;;) {
+      const { payload: { connectionStatus } } = yield take(UPDATE_CONNECTION_STATUS);
+
+      // We will ignore DISCONNECT actions until we connect
+
+      if (connectionStatus === ONLINE) {
+        // TODO: [P2] DirectLineJS should kill the connection when we unsubscribe
+        //       But currently in v3, DirectLineJS does not have this functionality
+        //       Thus, we need to call "end()" explicitly
+
+        return () => {
+          activitySubscription.unsubscribe();
+
+          try {
+            // TODO: [P3] DirectLineJS will throw an error saying "conversation ended", it seems redundant
+            directLine.end();
+          } catch (err) {}
+        };
+      } else if (connectionStatus === ENDED || connectionStatus === EXPIRED_TOKEN || connectionStatus === FAILED_TO_CONNECT) {
```

Comment:
I prefer using `Array.includes` for this:

```js
[ENDED, EXPIRED_TOKEN, FAILED_TO_CONNECT].includes(connectionStatus)
```

It makes it easier to see the statuses we care about when they are right next to each other.

Reply:
It looks expensive on the perf side (function call + object creation). Maybe I should line them up vertically so they are side-by-side too.

Reply:
Fixed by aligning vertically for readability.